contracts/token/ERC1155/ERC1155.sol

```diff
@@ -0,0 +1,234 @@
+        @param to Target address
+        @param id ID of the token type
+        @param value Transfer amount
+        @param data Data forwarded to `onERC1155Received` if `to` is a contract receiver
+    */
+    function safeTransferFrom(
+        address from,
+        address to,
+        uint256 id,
+        uint256 value,
+        bytes calldata data
+    )
+        external
+    {
+        require(to != address(0), "ERC1155: target address must be non-zero");
+        require(
+            from == msg.sender || _operatorApprovals[from][msg.sender] == true,
+            "ERC1155: need operator approval for 3rd party transfers"
+        );
+
+        _balances[id][from] = _balances[id][from].sub(value);
+        _balances[id][to] = value.add(_balances[id][to]);
+
+        emit TransferSingle(msg.sender, from, to, id, value);
+
+        _doSafeTransferAcceptanceCheck(msg.sender, from, to, id, value, data);
+    }
+
+    /**
+        @dev Transfers `values` amount(s) of `ids` from the `from` address to the
+        `to` address specified. Caller must be approved to manage the tokens being
+        transferred out of the `from` account. If `to` is a smart contract, will
+        call `onERC1155BatchReceived` on `to` and act appropriately.
+        @param from Source address
+        @param to Target address
+        @param ids IDs of each token type
+        @param values Transfer amounts per token type
+        @param data Data forwarded to `onERC1155Received` if `to` is a contract receiver
+    */
+    function safeBatchTransferFrom(
+        address from,
+        address to,
+        uint256[] calldata ids,
+        uint256[] calldata values,
+        bytes calldata data
+    )
+        external
+    {
+        require(ids.length == values.length, "ERC1155: IDs and values must have same lengths");
+        require(to != address(0), "ERC1155: target address must be non-zero");
+        require(
+            from == msg.sender || _operatorApprovals[from][msg.sender] == true,
+            "ERC1155: need operator approval for 3rd party transfers"
+        );
+
+        for (uint256 i = 0; i < ids.length; ++i) {
+            uint256 id = ids[i];
+            uint256 value = values[i];
+
+            _balances[id][from] = _balances[id][from].sub(value);
+            _balances[id][to] = value.add(_balances[id][to]);
+        }
+
+        emit TransferBatch(msg.sender, from, to, ids, values);
+
+        _doSafeBatchTransferAcceptanceCheck(msg.sender, from, to, ids, values, data);
+    }
+
+    /**
+     * @dev Internal function to mint an amount of a token with the given ID
+     * @param to The address that will own the minted token
+     * @param id ID of the token to be minted
+     * @param value Amount of the token to be minted
+     * @param data Data forwarded to `onERC1155Received` if `to` is a contract receiver
+     */
+    function _mint(address to, uint256 id, uint256 value, bytes memory data) internal {
```

Comment:
If possible, I'd like to punt on this, because this will require additional storage in the contract, and if implemented, probably _should_ prevent the use of counterfactual IDs (as you've mentioned below via reverting transactions on unminted IDs).

Since we (Gnosis) actually need to use counterfactual IDs in our implementation, we have not implemented this.

I can see how this may be useful for "game item" scenarios though: something that this standard was originally designed for, but because we don't want to exclude our own use case, I think this functionality would fit better in an derived contract.